crates/typst/src/layout/grid/lines.rs

```diff
@@ -0,0 +1,749 @@
+    pub position: LinePosition,
+}
+
+/// Indicates whether the line should be drawn before or after the track with
+/// its index. This is mostly only relevant when gutter is used, since, then,
+/// the position after a track is not the same as before the next
+/// non-gutter track.
+#[derive(PartialEq, Eq)]
+pub enum LinePosition {
+    /// The line should be drawn before its track (e.g. hline on top of a row).
+    Before,
+    /// The line should be drawn after its track (e.g. hline below a row).
+    After,
+}
+
+/// Generates the segments of lines that should be drawn alongside a certain
+/// axis in the grid, going through the given tracks (orthogonal to the lines).
+/// Each returned segment contains its stroke, its offset from the start, and
+/// its length.
+/// Accepts, as parameters, the index of the lines that should be produced
+/// (for example, the column at which vertical lines will be drawn); a list of
+/// user-specified lines with the same index (the `lines` parameter); whether
+/// the given index corresponds to the maximum index for the line's axis; and a
+/// function which returns the final stroke that should be used for each track
+/// the line goes through (its parameters are the grid, the index of the line
+/// to be drawn, the number of the track to draw at and the stroke of the user
+/// hline/vline override at this index to fold with, if any).
+/// Contiguous segments with the same stroke are joined together automatically.
+/// The function should return 'None' for positions at which the line would
+/// otherwise cross a merged cell (for example, a vline could cross a colspan),
+/// in which case a new segment should be drawn after the merged cell(s), even
+/// if it would have the same stroke as the previous one.
+/// Note that we assume that the tracks are sorted according to ascending
+/// number, and they must be iterable over pairs of (number, size). For
+/// vertical lines, for instance, 'tracks' would describe the rows in the
+/// current region, as pairs (row index, row height).
+pub(super) fn generate_line_segments<'grid, F, I>(
+    grid: &'grid CellGrid,
+    tracks: I,
+    index: usize,
+    lines: &'grid [Line],
+    is_max_index: bool,
+    line_stroke_at_track: F,
+) -> impl IntoIterator<Item = (Arc<Stroke<Abs>>, Abs, Abs)> + 'grid
+where
+    F: Fn(&CellGrid, usize, usize, Option<Arc<Stroke<Abs>>>) -> Option<Arc<Stroke<Abs>>>
+        + 'grid,
+    I: IntoIterator<Item = (usize, Abs)>,
+    I::IntoIter: 'grid,
+{
+    // The segment currently being drawn.
+    // It is extended for each consecutive track through which the line would
+    // be drawn with the same stroke.
+    // Starts as None to force us to create a new segment as soon as we find
+    // the first track through which we should draw.
+    let mut current_segment: Option<(Arc<Stroke<Abs>>, Abs, Abs)> = None;
+    // How far from the start (before the first track) have we gone so far.
```

Comment:
A blank line before a paragraph + let binding would make this a bit easier to read.